### PR TITLE
Update proguard README section to resolve #1261

### LINKED ISSUE #1261: Moshi ALWAYS fails to create adapter for ANY obfuscated enum class
Moshi always fails to create an adapter for any class that directly or indirectly references obfuscated enum class. The resulting stacktrace would look something like the following:
```
2020-10-23 14:52:51.063 8650-8650/dev.igokoro.moshi_enum E/AndroidRuntime: FATAL EXCEPTION: main
    Process: dev.igokoro.moshi_enum, PID: 8650
    java.lang.AssertionError: Missing field in c.a.a.a
        at b.b.a.r$k.<init>(:283)
        at b.b.a.r$b.a(:68)
        at b.b.a.q.f(:145)
        at b.b.a.c$a.b(:142)
        at b.b.a.c$a.a(:106)
        at b.b.a.q.f(:145)
        at b.b.a.q.e(:105)
        at b.b.a.q.c(:79)
        at dev.igokoro.moshi_enum.MainActivity.E(:27)
        at dev.igokoro.moshi_enum.MainActivity.onCreate(:15)
        at android.app.Activity.performCreate(Activity.java:8000)
        at android.app.Activity.performCreate(Activity.java:7984)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.NoSuchFieldException: A
        at java.lang.Class.getField(Class.java:1604)
        at b.b.a.r$k.<init>(:277)
        at b.b.a.r$b.a(:68) 
        at b.b.a.q.f(:145) 
        at b.b.a.c$a.b(:142) 
        at b.b.a.c$a.a(:106) 
        at b.b.a.q.f(:145) 
        at b.b.a.q.e(:105) 
        at b.b.a.q.c(:79) 
        at dev.igokoro.moshi_enum.MainActivity.E(:27) 
        at dev.igokoro.moshi_enum.MainActivity.onCreate(:15) 
        at android.app.Activity.performCreate(Activity.java:8000) 
        at android.app.Activity.performCreate(Activity.java:7984) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7656) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```
or deobfuscated
```
2020-10-23 14:52:51.063 8650-8650/dev.igokoro.moshi_enum E/AndroidRuntime: FATAL EXCEPTION: main
    Process: dev.igokoro.moshi_enum, PID: 8650
    java.lang.AssertionError: Missing field in c.a.a.a
        at com.squareup.moshi.StandardJsonAdapters$EnumJsonAdapter.<init>(:283)
        at com.squareup.moshi.StandardJsonAdapters$1.create(:68)
        at com.squareup.moshi.Moshi.adapter(:145)
        at com.squareup.moshi.ClassJsonAdapter$1.createFieldBindings(:142)
        at com.squareup.moshi.ClassJsonAdapter$1.create(:106)
        at com.squareup.moshi.Moshi.adapter(:145)
        at com.squareup.moshi.Moshi.adapter(:105)
        at com.squareup.moshi.Moshi.adapter(:79)
        at dev.igokoro.moshi_enum.MainActivity.java(:27)
        at dev.igokoro.moshi_enum.MainActivity.onCreate(:15)
        at android.app.Activity.performCreate(Activity.java:8000)
        at android.app.Activity.performCreate(Activity.java:7984)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.NoSuchFieldException: A
        at java.lang.Class.getField(Class.java:1604)
        at com.squareup.moshi.StandardJsonAdapters$EnumJsonAdapter.<init>(:277)
        at b.b.a.r$b.a(:68) 
        at b.b.a.q.f(:145) 
        at b.b.a.c$a.b(:142) 
        at b.b.a.c$a.a(:106) 
        at b.b.a.q.f(:145) 
        at b.b.a.q.e(:105) 
        at b.b.a.q.c(:79) 
        at dev.igokoro.moshi_enum.MainActivity.E(:27) 
        at dev.igokoro.moshi_enum.MainActivity.onCreate(:15) 
        at android.app.Activity.performCreate(Activity.java:8000) 
        at android.app.Activity.performCreate(Activity.java:7984) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7656) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```

Sample project to reproduce the issue: https://github.com/igokoro/moshi-obfuscated-enums

The root cause is that `EnumJsonAdapter` makes an assumption that entries from an array returned by `Class.getEnumConstants()` are always matching field names in the enum class. This does not hold true after the obfuscation that keeps enum constants unchanged but renames fields. (Looking into https://stackoverflow.com/questions/33189249/how-to-tell-proguard-to-keep-enum-constants-and-fields looks like such implementation originates at least from jaxb?)

The workaround is, of course, to exclude affected enums from obfuscation. 
But since the above will happen with any obfuscated enum:
* Is it possible to change `EnumJsonAdapter` to better tolerate obfuscated enums?
* Consider adding docs to the proguard section to describe the behavior and suggest an upfront solution/workaround. This can save a lot of time!

### PR DESCRIPTION
**Pull request by igokoro**
Update proguard README section to resolve #1261: make enums miss-handling harder by adding explicit instructions for enum classes that must be applied if obfuscation is used.

### CODE REVIEW

**Review by ZacSweers — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -706,6 +706,8 @@ R8 / ProGuard
 
 Moshi contains minimally required rules for its own internals to work without need for consumers to embed their own. However if you are using reflective serialization and R8 or ProGuard, you must add keep rules in your proguard configuration file for your reflectively serialized classes.
 
+**Note on enums:** any enum classes (java or kotlin) used with moshi must be annotated with `@JsonClass(generateAdapter = false)`. Not annotating the class will result in `NoSuchFieldException` being thrown at adapter creation stage.
```
Comment (ZacSweers) [inline README.md:709]:
```suggestion
**Note on enums:** any enum classes (Java or Kotlin) used with Moshi must be annotated with `@JsonClass(generateAdapter = false)`. Not annotating the class can result in `NoSuchFieldException` at runtime if the member represented in JSON isn't present in the enum.
```

**Review by ZacSweers — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (igokoro) [reply README.md:709]: @ZacSweers, thanks for your input! I'm OK accepting your suggestions - just want to make sure we're on the same page on the actual issue:
the issue I was describing in #1261 will always happen for any obfuscated enum before Moshi even gets to reading/writing JSON. The exception is thrown at adapter creation time in `EnumJsonAdapter` constructor. `@JsonClass(generateAdapter = false)` fixes this as proguard/R8 is instructed to leave enum fields untouched.

The changes you suggested imply a slightly different perspective - outcome depends on the JSON contents, which I believe is not accurate in this case.

Comment (igokoro) [reply README.md:709]: @ZacSweers, I wonder if you had a chance to look through my reply?

Comment (ZacSweers) [reply README.md:709]: It does matter and it's a little more detailed than your description. I'd rather overcommunicate the risks here

Comment (igokoro) [reply README.md:709]: So, 2 things:
1. I'm not sure how I feel about this interaction... I was asked to submit a PR upon opening a defect which I did. Now I'm getting vague feedback that is very easy to interpret like "just accept my change and let's be done with this". I don't want to look negative and I totally understand and accept that you may have other more important things going on in your day-to-day life. I guess I'd just like to see an actual discussion if I'm asked to be part of that discussion. The discussion itself can wait until a better time.
2. > I'd rather overcommunicate the risks here
Well, I don't think the message you suggested is overcommunicating. If I understand correctly, it covers cases where proguard renames enums, and those do not match what is in JSON anymore. But the thing is - any obfuscated enum present anywhere in the object tree fed to `Moshi` will cause the adapter creation (`moshi.adapter(...)`) to fail with `NoSuchFieldException`. 
To put this in context, the regular android application would be failing at `Retrofit` service method invocation time. There will be no network request executed, no JSON received. For that matter, contents of the received JSON are also irrelevant in this specific context... Please let me know if my understanding here is not correct.

**Comment by thegarlynch** [comment]
This still not solved for inner enum class

**Comment by ZacSweers** [comment]
Then suggest a fix

**Comment by ZacSweers** [comment]
Closing for now as this has gone stale

**Pull request closed without merging**